### ISSUE TITLE
Generic MyClass<T> without property of type T should raise an error

### ISSUE THREAD

**Issue by Speedy37** [Microsoft/TypeScript#10251]
**TypeScript Version:** 1.8.4 / 2.0.0 / 2.1.0-dev.20160720

**Code**

```ts
declare class MyClass<T> {
    then(): MyClass<T>;
}
var test: MyClass<boolean> = new MyClass<string>();
```

**Expected behavior:**

```
Type 'MyClass<string>' is not assignable to type 'MyClass<boolean>'.
  Type 'string' is not assignable to type 'boolean'.
```

**Actual behavior:** 

No error

**Workaround**

```ts
declare class MyClass<T> {
    then(): MyClass<T>;
    _(): T;
}
var test: MyClass<boolean> = new MyClass<string>();
```

**Comment by RyanCavanaugh** [comment]
People do this *on purpose* a lot despite it being inadvisable.

The other problem, as shown in your example, is that we have to perform potentially infinite expansion to be sure that there's no instantiated `T` properties anywhere. We also can't just do it syntactically because there might be interface augmentations elsewhere.
